binaries/opae.io/opae/io/utils.py

```diff
@@ -448,54 +423,63 @@ def get_conf():
     return {}
 
 
+def get_conf_devices(data):
+    conf_ids = {}
+    for k, v in data.items():
+        try:
+            ids = k.split(':')
+        except:
+            print(f'error with vendor/device: {k}')
+        else:
+            conf_ids[pci.pci_id(*[int(i, 16) for i in ids])] = v
+    return conf_ids
+
+
+
 def lsfpga(**kwargs):
-    vendor = kwargs.get('vendor')
-    device = kwargs.get('device')
-    driver = kwargs.get('driver')
+    _all = kwargs.pop('all', False)
     device_ids = dict(fpga_devices)
+    use_class = kwargs.pop('system_class', False)
     conf = get_conf()
-    conf_ids = {}
     if conf:
-        for k, v in conf.get('fpga_devices', {}).items():
-            try:
-                vstr, dstr = k.split(':')
-            except:
-                print('error with vendor/device: {}'.format(k))
-            else:
-                conf_ids[(int(vstr, 16), int(dstr, 16))] = v
-
+        conf_ids = get_conf_devices(conf.get('fpga_devices', {}))
         if conf.get('override', False):
             device_ids.update(conf_ids)
         else:
             for k, v in conf_ids.items():
                 if k not in device_ids:
                     device_ids[k] = v
 
-    if kwargs.get('all'):
-        device_ids = {}
-
-    for pcidir in Path('/sys/bus/pci/devices').glob('*'):
-        vid = read_attr(pcidir, 'vendor')
-        did = read_attr(pcidir, 'device')
-        drv = read_link(pcidir, 'driver')
-        if drv:
-            drv = drv.stem
-        if vid is None or did is None:
-            continue
-        if vendor and int(vid, 16) != int(vendor, 16):
-            continue
-        if device and int(did, 16) != int(device, 16):
-            continue
-        id_tuple = (int(vid, 0), int(did, 0))
-        if not device_ids or id_tuple in device_ids:
-            if not driver or (driver and drv and driver == drv):
-                yield (pcidir.parts[-1], vid, did, device_ids.get(id_tuple, ''), drv)
+
+    def filter_fn(d: pci.device):
+        for k,v in kwargs.items():
```

Comment:
Is kwargs a parameter here?  Should the function signature account for it?
```python
def filter_fn(d: pci.device, **kwargs):
```

Reply:
no, `filter_fn` is a closure and defined inside of the outer function which means its scope includes the outer function.

Reply:
Perhaps a comment that clarifies what's going on?